conans/client/tools/scm.py

```diff
@@ -89,23 +99,45 @@ def run(self, command):
         command = self._configure_ssl_verify + command
         return super(Git, self).run(command)
 
-    def clone(self, url, branch=None, args=""):
+    def clone(self, url, branch=None, args="", shallow=False):
+        """
+        :param url: repository remote URL to clone from (e.g. https, git or local)
+        :param branch: actually, can be any valid git ref expression like,
+        - None, use default branch, usually it's "master"
+        - branch name
+        - tag name
+        - revision sha256
+        - expression like HEAD~1
+        :param args: additional arguments to be passed to the git command (e.g. config args)
+        :param shallow:
+        :return: output of the clone command
+        """
+        # TODO: rename "branch" -> "element" in Conan 2.0
         url = self.get_url_with_credentials(url)
         if os.path.exists(url):
             url = url.replace("\\", "/")  # Windows local directory
-        if os.path.exists(self.folder) and os.listdir(self.folder):
+        if (os.path.exists(self.folder) and os.listdir(self.folder)) or (shallow and branch):
```

Comment:
What if... `self.folder` doesn't exist and `shallow and branch` evaluates to `True`? The first `git init` is going to fail...

Reply:
do you think if it should have `os.makedirs` then?

Reply:
No, but those `if ... and ...` where there for a reason. Why do we need now the extra `or (shallow and branch)`?

Reply:
for shallow, we can't do `git clone <sha256>`, so we need to use fetch-based approach, even if folder doesn't exist. so I guess `makedirs` is harmless here, right?

Reply:
I don't understand, here we are doing (in this PR) `git clone --branch <branch> --depth 1`. 

I still don't know why if `shallow and branch` we need to run the same sources as if the `self.folder` exists.

Reply:
we never do `git clone --branch <branch> --depth 1`, either `git clone --depth 1`, or `git clone --branch <branch>`. the reason, is that we want to be able to shallow clone of the specific revision, which cannot be done via `git clone` command, only via `git fetch`

Reply:
Ok, you are right, because if `shallow and branch` we never enter the `else`... 

IMHO the line `if (os.path.exists(self.folder) and os.listdir(self.folder)) or (shallow and branch):` is mixing things that have nothing to do with each other, I find it a little bit obfuscated (that the reason I was wrong in my previous comment). Maybe it is worth reorganizing the `if/else` then and it will result in cleaner code, don't you think so?

Reply:
do you have a suggestion how to reorganise?

Reply:
Nothing special, maybe we just have `2x2` branches here, I need to see the actual git commands to know if it is easier to understand or not.

Reply:
refactored the condition

Reply:
Now it looks much better (even better than before). Need to review it all again.

Thanks